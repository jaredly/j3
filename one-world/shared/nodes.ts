//

import { splitGraphemes } from '../../src/parse/splitGraphemes';
import { filterNulls } from '../../web/custom/old-stuff/filterNulls';
import { selectNode } from '../client/selectNode';
import { isCollection } from '../client/TextEdit/actions';
import { RenderInfo } from './renderables';
import { NodeSelection } from './state';

export type Loc = Array<[string, number]>;

export type PathRoot = {
    type: 'doc-node';
    doc: string;
    ids: number[];
    toplevel: string;
}; // another option would be like, "search-result" or "pin" or something

export type Path = {
    root: PathRoot;
    children: number[];
};

export const pathWithChildren = (path: Path, ...children: number[]) => ({
    ...path,
    children: path.children.concat(children),
});

export const parentPath = (path: Path) => ({
    ...path,
    children: path.children.slice(0, -1),
});

export const serializePath = (path: Path) => {
    return `${path.root.doc}#${path.root.ids.join('.')}#${path.children.join(
        '.',
    )}`;
};

export type Selection =
    | { type: 'start' | 'end' | 'inside' }
    | { type: 'subtext'; at: number }
    | { type: 'rich-text'; sel: any };

export type Cursor = {
    path: Path;
    selection: Selection;
};

/*
-------- Rich Text ----------
*/

export type Style = {
    fontWeight?: number | string;
    fontFamily?: string;
    fontStyle?: string;
    textDecoration?: string;
    background?: { r: number; g: number; b: number } | false;
    border?: string;
    outline?: string;
    color?: { r: number; g: number; b: number } | false;
};

type StyleSpan = {
    style: Style;
    start: number;
    end: number;
};

// export type RichBlock =
//     | { type: 'r:h'; level: number; contents: RichInline[] }
//     | { type: 'r:p'; contents: RichInline[] }
//     | { type: 'r:l'; ordered: boolean; contents: RichBlock[] }
//     | { type: 'r:checks'; contents: { check: boolean; block: RichBlock[] }[] }
//     | { type: 'r:opts'; which: number; contents: RichBlock[] }
//     | { type: 'r:indent'; quote: boolean; contents: RichBlock[] }
//     | { type: 'r:table'; rows: RichBlock[][] }
//     | { type: 'r:hr' };

// export type RichInline =
//     | { type: 'text'; style: Style; text: string }
//     | { type: 'link'; style: Style; text: string; url: string }
//     | { type: 'image'; url: string; style: Style }
//     | { type: 'embed'; node: Node; format?: FormatMap };

// export type FormatMap = Record<string, RenderInfo>;

//////// ALTERNATIVE TWO

/*

So the neat thing about this, is that rich-text nodes and sub-nodes
are ~no different from other syntax nodes.

ooh so the ... I mean ...
Yeah I like the unity of it.


*/

export type InlineSpan =
    | {
          type: 'text';
          text: string;
          style: Style;
      }
    | { type: 'link'; text: string; link: string; style: Style }
    | { type: 'embed'; item: number };

// Rich Paragraph(?)
export type RichInline = {
    type: 'rich-inline';
    spans: InlineSpan[];
    kind?: InlineKind;
    loc: number;
};

export type RichBlock = {
    type: 'rich-block';
    items: number[];
    loc: number;
    kind?: BlockKind;
    style: Style;
};

// Do we want footnotes? Or "hover for more info"s?
export type InlineSpanT<Loc> =
    | {
          type: 'text';
          text: string;
          style: Style;
      }
    | { type: 'link'; text: string; link: string; style: Style }
    | { type: 'embed'; item: RecNodeT<Loc> };

// So, when editing, default is to start with a rich-inline,
// but if you start needing extra formatting, we wrap in a
// rich-block.
export type RichInlineT<Loc> = {
    type: 'rich-inline';
    spans: InlineSpanT<Loc>[];
    kind?: InlineKind;
    loc: Loc;
};

export type RichBlockT<Loc> = {
    type: 'rich-block';
    items: RecNodeT<Loc>[];
    loc: Loc;
    kind?: BlockKind;
    style: Style;
};

type InlineKind = { type: 'header'; level: number };

type BlockKind =
    | { type: 'list'; ordered: boolean }
    | { type: 'checks'; checked: Record<number, boolean> }
    | { type: 'opts'; which: number }
    | { type: 'indent'; quote: boolean }
    | { type: 'callout'; vibe: 'info' | 'warning' | 'error' };

/*
-------- Sexp Syntax Nodes ----------
*/

type Id<Loc> =
    // id for identifier. "blank" === empty id
    {
        type: 'id';
        text: string;
        loc: Loc;
        ref?: { toplevel: string; loc: number };
    };

export type Node =
    | Id<number>
    | RichInline
    | RichBlock
    | { type: 'list' | 'array' | 'record'; items: number[]; loc: number }
    | {
          type: 'string';
          tag: number;
          first: string;
          templates: { expr: number; suffix: string }[];
          loc: number;
      }
    | { type: 'comment' | 'spread'; contents: number; loc: number }
    | { type: 'annot'; contents: number; annot: number; loc: number }
    | { type: 'record-access'; target: number; items: number[]; loc: number };
// doooo I want to be embedding some embeds? I kinda want to leave open the option.
// | { type: 'rich-text'; contents: any; loc: number; embeds: number[] };
// | {
//       type: 'raw-code';
//       lang: string;
//       raw: string;
//       loc: number;
//       embeds: number[];
//   };

export type RecNode = RecNodeT<Loc>;

export type RecNodeT<Loc> =
    | Id<Loc>
    | RichInlineT<Loc>
    | RichBlockT<Loc>
    | { type: 'list' | 'array' | 'record'; items: RecNodeT<Loc>[]; loc: Loc }
    | {
          type: 'string';
          tag: RecNodeT<Loc>;
          first: string;
          templates: { expr: RecNodeT<Loc>; suffix: string }[];
          loc: Loc;
      }
    | { type: 'comment' | 'spread'; contents: RecNodeT<Loc>; loc: Loc }
    | { type: 'annot'; contents: RecNodeT<Loc>; annot: RecNodeT<Loc>; loc: Loc }
    | {
          type: 'record-access';
          target: RecNodeT<Loc>;
          items: RecNodeT<Loc>[];
          loc: Loc;
      };

export type Nodes = Record<number, Node>;

export const inFromEnd = (
    node: Node,
    path: Path,
    nodes: Nodes,
): void | NodeSelection => {
    if (node.type === 'string') {
        return {
            type: 'string',
            path,
            pathKey: serializePath(path),
            cursor: {
                part: node.templates.length,
                char: splitGraphemes(
                    node.templates.length === 0
                        ? node.first
                        : node.templates[node.templates.length - 1].suffix,
                ).length,
            },
        };
    }
    const children = childLocs(node);
    if (!children.length) {
        return;
    }
    const loc = children[children.length - 1];
    return selectNode(nodes[loc], pathWithChildren(path, loc), 'end', nodes);
};

export const inFromStart = (
    node: Node,
    path: Path,
    nodes: Nodes,
): void | NodeSelection => {
    if (node.type === 'string') {
        return {
            type: 'string',
            cursor: { part: 0, char: 0 },
            path,
            pathKey: serializePath(path),
        };
    }
    const children = childLocs(node);
    if (children.length === 0) {
        return;
    }
    const loc = children[0];
    return selectNode(nodes[loc], pathWithChildren(path, loc), 'start', nodes);
};

export const firstAtom = (path: Path, nodes: Nodes): Path => {
    const loc = path.children[path.children.length - 1];
    const node = nodes[loc];
    if (node.type === 'string') return path;
    const children = childLocs(node);
    if (!children.length) {
        return path;
    }
    return firstAtom(
        { ...path, children: path.children.concat([children[0]]) },
        nodes,
    );
};

export const lastAtom = (path: Path, nodes: Nodes): Path => {
    const loc = path.children[path.children.length - 1];
    const node = nodes[loc];
    if (node.type === 'string') return path;
    const children = childLocs(node);
    if (!children.length) {
        return path;
    }
    return lastAtom(
        {
            ...path,
            children: path.children.concat([children[children.length - 1]]),
        },
        nodes,
    );
};

export const nextAtom = (path: Path, nodes: Nodes): Path | void => {
    if (path.children.length < 2) return; // ok um
    const loc = path.children[path.children.length - 1];
    const ploc = path.children[path.children.length - 2];
    const parent = nodes[ploc];
    if (parent.type === 'string') return;
    const children = childLocs(parent);
    if (!children.length) {
        return;
    }

    const idx = children.indexOf(loc);
    if (idx === children.length - 1) {
        return nextAtom(parentPath(path), nodes);
    }
    return firstAtom(
        pathWithChildren(parentPath(path), children[idx + 1]),
        nodes,
    );
};

export const prevAtom = (path: Path, nodes: Nodes): Path | void => {
    if (path.children.length < 2) return; // ok um
    const loc = path.children[path.children.length - 1];
    const ploc = path.children[path.children.length - 2];
    const parent = nodes[ploc];
    if (parent.type === 'string') return;
    const children = childLocs(parent);
    if (!children.length) {
        return;
    }

    const idx = children.indexOf(loc);
    if (idx === 0) {
        return prevAtom(parentPath(path), nodes);
    }
    return lastAtom(
        pathWithChildren(parentPath(path), children[idx - 1]),
        nodes,
    );
};

export const toTheRight = (
    parent: Node,
    cloc: number,
    path: Path,
    nodes: Nodes,
): void | NodeSelection => {
    if (parent.type === 'string') {
        const ppath = path; //parentPath(path);
        const pathKey = serializePath(ppath);
        if (cloc === parent.tag) {
            return {
                type: 'string',
                path: ppath,
                pathKey,
                cursor: { part: 0, char: 0 },
            };
        }
        const at = parent.templates.findIndex((t) => t.expr === cloc);
        if (at === -1) return;
        return {
            type: 'string',
            path: ppath,
            pathKey,
            cursor: { part: at + 1, char: 0 },
        };
    }
    const children = childLocs(parent);
    const idx = children.indexOf(cloc);
    if (idx === children.length - 1) {
        return selectNode(parent, path, 'end', nodes);
    }
    const loc = children[idx + 1];
    return selectNode(nodes[loc], pathWithChildren(path, loc), 'start', nodes);
};

export const toTheLeft = (
    parent: Node,
    cloc: number,
    path: Path,
    nodes: Nodes,
): void | NodeSelection => {
    if (parent.type === 'string') {
        if (cloc === parent.tag) {
            const gparent = path.children[path.children.length - 2];
            return toTheLeft(
                nodes[gparent],
                parent.loc,
                parentPath(path),
                nodes,
            );
        }
        const ppath = path; // parentPath(path);
        const pathKey = serializePath(ppath);
        const at = parent.templates.findIndex((t) => t.expr === cloc);
        if (at === -1) return;
        return {
            type: 'string',
            path: ppath,
            pathKey,
            cursor: {
                part: at,
                char: splitGraphemes(
                    at === 0 ? parent.first : parent.templates[at - 1].suffix,
                ).length,
            },
        };
    }
    const children = childLocs(parent);
    const idx = children.indexOf(cloc);
    if (idx === 0) {
        return selectNode(parent, path, 'start', nodes);
    }
    const loc = children[idx - 1];
    return selectNode(nodes[loc], pathWithChildren(path, loc), 'end', nodes);
};

export const childLocs = (node: Node) => {
    switch (node.type) {
        case 'id':
            return [];
        case 'rich-inline':
            return node.spans
                .map((s) => (s.type === 'embed' ? s.item : null))
                .filter(filterNulls);
        case 'rich-block':
            return node.items;
        case 'list':
        case 'array':
        case 'record':
            return node.items;
        case 'comment':
        case 'spread':
            return [node.contents];
        case 'annot':
            return [node.contents, node.annot];
        case 'record-access':
            return [node.target, ...node.items];
        case 'string':
            return [node.tag, ...node.templates.flatMap((t) => t.expr)];
    }
};

export const fromMap = <Loc>(
    getLoc: (loc: number) => Loc,
    id: number,
    nodes: Nodes,
    paths?: { children: number[]; map: Record<number, number[]> },
): RecNodeT<Loc> => {
    if (paths) {
        paths.map[id] = paths.children;
    }
    const sub = paths
        ? { ...paths, children: paths.children.concat([id]) }
        : undefined;
    const node = nodes[id];
    const loc = getLoc(node.loc); // : [string, number][] = [[top, node.loc]];
    switch (node.type) {
        case 'id':
            return { ...node, loc };
        case 'rich-block':
        case 'list':
        case 'array':
        case 'record':
            return {
                ...node,
                loc,
                items: node.items.map((n) => fromMap(getLoc, n, nodes, sub)),
            };
        case 'comment':
        case 'spread':
            return {
                ...node,
                loc,
                contents: fromMap(getLoc, node.contents, nodes, sub),
            };
        case 'annot':
            return {
                ...node,
                loc,
                contents: fromMap(getLoc, node.contents, nodes, sub),
                annot: fromMap(getLoc, node.annot, nodes, sub),
            };
        case 'record-access':
            return {
                ...node,
                loc,
                target: fromMap(getLoc, node.target, nodes, sub),
                items: node.items.map((n) => fromMap(getLoc, n, nodes, sub)),
            };
        case 'rich-inline':
            return {
                ...node,
                loc,
                spans: node.spans.map((t) =>
                    t.type === 'embed'
                        ? {
                              type: 'embed',
                              item: fromMap(getLoc, t.item, nodes, sub),
                          }
                        : t,
                ),
            };
        case 'string':
            return {
                ...node,
                loc,
                tag: fromMap(getLoc, node.tag, nodes, sub),
                templates: node.templates.map((t) => ({
                    expr: fromMap(getLoc, t.expr, nodes, sub),
                    suffix: t.suffix,
                })),
            };
    }
};

const foldNode = <V>(v: V, node: RecNode, f: (v: V, node: RecNode) => V): V => {
    switch (node.type) {
        case 'id':
        case 'rich-inline':
            return f(v, node);
        case 'list':
        case 'array':
        case 'record':
        case 'rich-block':
            return node.items.reduce(
                (v, node) => foldNode(v, node, f),
                f(v, node),
            );
        case 'comment':
        case 'spread':
            return foldNode(f(v, node), node.contents, f);
        case 'annot':
            return foldNode(
                foldNode(f(v, node), node.contents, f),
                node.annot,
                f,
            );
        case 'record-access':
            return node.items.reduce(
                (v, node) => foldNode(v, node, f),
                foldNode(f(v, node), node.target, f),
            );
        case 'string':
            return node.templates.reduce(
                (v, { expr }) => foldNode(v, expr, f),
                f(v, node),
            );
    }
};

const maxLoc = (node: RecNode) => {
    return foldNode(0, node, (max, node) =>
        node.loc.length ? Math.max(max, node.loc[node.loc.length - 1][1]) : max,
    );
};

// Gotta have a way to determine

const getLoc = (node: RecNode, nodes: Nodes, idx: { next: number }) => {
    if (!node.loc.length) return idx.next++;
    const loc = node.loc[node.loc.length - 1][1];
    if (loc in nodes) {
        return idx.next++;
    }
    // @ts-ignore this will get overwritten in a sec
    nodes[loc] = null;
    return loc;
};

export const toMap = (node: RecNode, nodes: Nodes): number => {
    const idx = { next: maxLoc(node) + 1 };
    return toMapInner(node, [], nodes, (node, nodes) =>
        getLoc(node, nodes, idx),
    );
};

export const toMapInner = <T>(
    node: RecNodeT<T>,
    path: number[],
    nodes: Nodes,
    getLoc: (node: RecNodeT<T>, nodes: Nodes, path: number[]) => number,
    // idx: { next: number },
): number => {
    const loc = getLoc(node, nodes, path);
    nodes[loc] = fromRec(node, path.concat([loc]), loc, nodes, getLoc);
    return loc;
};

const fromRec = <T>(
    node: RecNodeT<T>,
    path: number[],
    loc: number,
    nodes: Nodes,
    getLoc: (node: RecNodeT<T>, nodes: Nodes, path: number[]) => number,
): Node => {
    switch (node.type) {
        case 'id':
            return { ...node, loc };
        case 'list':
        case 'array':
        case 'record':
        case 'rich-block':
            return {
                ...node,
                loc,
                items: node.items.map((n) =>
                    toMapInner(n, path, nodes, getLoc),
                ),
            };
        case 'comment':
        case 'spread':
            return {
                ...node,
                loc,
                contents: toMapInner(node.contents, path, nodes, getLoc),
            };
        case 'annot':
            return {
                ...node,
                loc,
                contents: toMapInner(node.contents, path, nodes, getLoc),
                annot: toMapInner(node.annot, path, nodes, getLoc),
            };
        case 'record-access':
            return {
                ...node,
                loc,
                target: toMapInner(node.target, path, nodes, getLoc),
                items: node.items.map((n) =>
                    toMapInner(n, path, nodes, getLoc),
                ),
            };
        case 'rich-inline':
            return {
                ...node,
                loc,
                spans: node.spans.map((t) =>
                    t.type === 'embed'
                        ? {
                              item: toMapInner(t.item, path, nodes, getLoc),
                              type: 'embed',
                          }
                        : t,
                ),
            };
        case 'string':
            return {
                ...node,
                loc,
                tag: toMapInner(node.tag, path, nodes, getLoc),
                templates: node.templates.map((t) => ({
                    expr: toMapInner(t.expr, path, nodes, getLoc),
                    suffix: t.suffix,
                })),
            };
    }
};
