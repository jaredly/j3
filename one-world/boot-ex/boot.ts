import { RecNode } from '../shared/nodes';
import { RenderInfo } from '../shared/renderables';

type Fmt = { loc: number; info: RenderInfo };
type Usage = { decl: number; usage?: number };
type ParseError = { loc: number; msg: string };
type Autocomplete = {
    kind: string;
    options: string[];
};

// ALTERNATIVE:
// format takes a "parsed form" and returns the list of formats about it.
// which we could also do with the normal one, to be fair

type Ctx = {
    cursor?: { loc: number; autocomplete: [] };
    errors: ParseError[];
    fmt: Fmt[];
    usages: Usage[];
    exports: { loc: number; kind: string }[];
};

const parse = (node: RecNode, ctx: Ctx) => {
    // do the thing
    //
    // parse gives us
    // - the node
    // - any accumilated juices .. .I mean errors
    // - formatting infos
    // - usages (for "rename symbol" and highlight-under-cursor)
    // - exports (locs of identifiers that are "exports", along with a "kind" (e.g. value, type, effect))
    // - autocomplete infos if we have a cursor here
};

// DO I NEED
// to have a "parse expression" form?
// like, idk I probably don't? As in, anywhere I might
// want to parse an expression (cough, plugins)
// I could ... hrm
// ok so the thing I could do is:
// -> have a `asExpr(toplevel) : Option(expr)`
// which, honestly, might not be that bad.
// like, better to only have one entrypoint for the parsing, I think?

// Do I actually need this?
// hmmm maybe? or is it just a crutch?
// idk. Might be good to think on this more.
const fromNode = (a: RecNode) => a;
const toNode = (a: RecNode) => a;

const asExpr = (t: any) => {
    if (t.type === 'texpr') return { type: 'some', 0: t[0] };
    return { type: 'none' };
};

// ------------------ Compiler -------------------
// ok do we need compileExpr?
// kinda seems like we do.
const compileTop = (
    abc: any,
    typeInfo: any,
    trace: { top: boolean; nodes: Record<number, { fmt?: string }> },
): string => {
    return 'lol';
};

// -------------------- Runtime -------------------
// the stuff that when evaled returns a map of things.
// these get plopped onto `$env`.
const builtins = '';

// -------------------- Type Inference ---------------

/*
init(): tenv with builtins
^ NOTE: should ... the init ... include builtins? seems like it.
// or I could have a separate thing that is "builtins". I mean,
// why bother with that. Yeah let's wait on that until/if I need it.
infer-top(toplevel, tenv):
    - tenv
    - (option type)  ; the type to display over this toplevel
    - map[loc, type] ; for displaying type information for globals when autocompleting and such. and inthe naemspce listing
        ^ Q:Q:Q: should the type here already be a `cst`?
        ^ should that CST also include formatting infos?
        ^ one formatting info might be "links to [toplevel/loc]"
    - code gen data
    - inference errors maybe
    - types for locs
add-tenv(old, new): tenv
type-to-cst(type): cst, fmts






*/
