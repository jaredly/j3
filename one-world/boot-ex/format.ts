// It's a parser, that only does formatting!

import { Layout } from '../shared/IR/intermediate';
import { Loc, RecNode, Style } from '../shared/nodes';

export const parse = (node: RecNode) => {
    const ctx: Ctx = { layouts: {}, styles: {} };

    const result = _parse(node, ctx);
    return { result, ...ctx };
};

// type Fmt = { loc: number; info: RenderInfo };
type Ctx = {
    // cursor?: { loc: number; autocomplete?: Autocomplete };
    // errors: ParseError[];
    // fmt: Fmt[];
    layouts: Record<number, Layout>;
    styles: Record<number, Style>;
    // usages: Usage[];
    // exports: { loc: number; kind: string }[];
};

// const tights: Record<string, Layout> = {
//     if: {type: 'vert', layout: {tightFirst: 2, indent: 2}},
//     '->': 1,
//     match: 1,
// }

const getLoc = (l: Loc) => l[l.length - 1][1];

const _parse = (node: RecNode, ctx: Ctx) => {
    switch (node.type) {
        case 'list':
            if (node.items.length > 0) {
                const first = node.items[0];
                if (first.type === 'id') {
                    switch (first.text) {
                        case 'if':
                            ctx.layouts[getLoc(node.loc)] = {
                                type: 'vert',
                                layout: { tightFirst: 2, indent: 2 },
                            };
                            break;
                        case 'let':
                            ctx.layouts[getLoc(node.loc)] = {
                                type: 'vert',
                                layout: { tightFirst: 2, indent: 2 },
                            };
                            if (node.items[1]?.type === 'array') {
                                ctx.layouts[getLoc(node.items[1].loc)] = {
                                    type: 'vert',
                                    layout: {
                                        tightFirst: 0,
                                        indent: 0,
                                        pairs: [],
                                    },
                                };
                            }
                            break;
                        case 'defn':
                            ctx.layouts[getLoc(node.loc)] = {
                                type: 'vert',
                                layout: { tightFirst: 3, indent: 4 },
                            };
                            break;
                    }
                }
            }
        case 'array':
            node.items.map((p) => _parse(p, ctx));
            return;
        case 'record':
            ctx.layouts[getLoc(node.loc)] = {
                type: 'vert',
                layout: { tightFirst: 0, indent: 0, pairs: [] },
            };
            node.items.map((p) => _parse(p, ctx));
            return;
        // case 'id':
        //     if (node.text)
    }
};
