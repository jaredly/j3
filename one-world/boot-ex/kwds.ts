// builtin forms, idk. autocomplete kind = kwd

import { RecNodeT } from '../shared/nodes';

type Auto = {
    text: string;
    toplevel?: boolean;
    // Sooo I kinda want there to be able to be multiple templates to choose from?
    // idk like (fn x) and (fn [a] x) right?
    // idk if it's too niche.
    // like macros can totally admit muiltiple forms.
    // ok I'll do it now.
    templates: { template: RecNodeT<boolean>[]; docs: string }[];
};

const place = (text: string, focus = false): RecNodeT<boolean> => ({
    type: 'id',
    loc: focus,
    text: '',
    ref: { type: 'placeholder', text },
});

export const kwds: Auto[] = [
    {
        text: 'let',
        templates: [
            {
                template: [
                    {
                        type: 'table',
                        loc: false,
                        rows: [[place('pattern', true), place('value')]],
                    },
                    place('body'),
                ],
                docs: 'Locally bind variables',
            },
        ],
    },
    {
        text: 'match',
        templates: [
            {
                template: [
                    place('value', true),
                    {
                        type: 'table',
                        loc: false,
                        rows: [[place('pattern', true), place('value')]],
                    },
                ],
                docs: 'Destructure ADTs',
            },
        ],
    },
    {
        text: 'if',
        templates: [
            { template: [place('cond'), place('yes'), place('no')], docs: '' },
        ],
    },
    {
        text: '->',
        templates: [
            {
                template: [place('target'), place('transform')],
                docs: 'first argument chain',
            },
        ],
    },
    {
        text: '->>',
        templates: [
            {
                template: [place('target'), place('transform')],
                docs: 'last argument chain',
            },
        ],
    },
    {
        text: 'fn',
        templates: [
            {
                template: [
                    { type: 'array', loc: false, items: [place('pattern')] },
                    place('body'),
                ],
                docs: 'lambda function',
            },
        ],
    },
    {
        text: 'test',
        templates: [
            {
                template: [
                    place('function'),
                    {
                        type: 'table',
                        loc: false,
                        rows: [[place('input'), place('output')]],
                    },
                ],
                docs: 'fixture tests',
            },
        ],
    },
    {
        text: 'def',
        toplevel: true,
        templates: [
            {
                template: [place('name', true), place('value')],
                docs: 'global definition',
            },
        ],
    },
    {
        text: 'defn',
        toplevel: true,
        templates: [
            {
                template: [
                    place('name', true),
                    { type: 'array', items: [place('pattern')], loc: false },
                    place('body'),
                ],
                docs: 'global function definition',
            },
        ],
    },
    {
        text: 'deftype',
        toplevel: true,
        templates: [
            {
                template: [
                    place('name', true),
                    {
                        type: 'list',
                        loc: false,
                        items: [place('constructor'), place('argtype')],
                    },
                ],
                docs: 'global enum type definition',
            },
        ],
    },
];
