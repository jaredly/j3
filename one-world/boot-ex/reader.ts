import { IDRef, Loc, RecNode } from '../shared/nodes';

type RNode = RecNode;

export const reader = (
    text: string,
    top: string = 'test-top',
    globals: Record<string, IDRef> = {},
): RNode | void => {
    let i = 0;
    const pairs = { '[': ']', '{': '}', '(': ')' };
    const skipWhite = () => {
        while (i < text.length && text[i].match(/\s/)) i++;
        return i >= text.length;
    };
    const l = (num: number): Loc => [[top, num]];

    const readString = (): Extract<RNode, { type: 'string' }> | void => {
        const loc = i;
        i++;
        let first = '';
        let current = '';
        let templates: { suffix: string; expr: RNode; loc: number }[] = [];

        const next = () => {
            if (!templates.length) {
                first = current;
            } else {
                templates[templates.length - 1].suffix = current;
            }
            current = '';
        };

        for (; i < text.length; i++) {
            if (text[i] === '"') {
                i++;
                break;
            }
            if (text[i] === '\\') {
                current += text[i];
                i++;
            }
            if (text[i] === '$' && text[i + 1] === '{') {
                next();
                i += 2;
                const expr = read();
                if (!expr) return;
                templates.push({ expr, suffix: '', loc: i });
                if (text[i] !== '}')
                    throw new Error('unmatched } in template string');
                continue;
            }
            current += text[i];
            // console.log('a', current, text[i])
        }
        next();
        return {
            first,
            templates,
            type: 'string',
            loc: l(loc),
            tag: { type: 'id', text: '', loc: l(loc) },
        };
    };

    const stops = `|"]}) \t\n`;
    const read = (): RNode | void => {
        if (skipWhite()) return;
        if (text[i] === '(' && text[i + 1] === '|') {
            const start = i;
            i += 2;
            const rows: RNode[][] = [[]];
            while (true) {
                if (text[i] === '\n') {
                    rows.push([]);
                    i++;
                }
                if (skipWhite()) return;
                if (text[i] === '|' && text[i + 1] === ')') {
                    break;
                }
                if (text[i] === '|') {
                    i++;
                }
                const next = read();
                if (!next) break;
                rows[rows.length - 1].push(next);
            }
            return {
                type: 'table',
                kind: '(',
                rows,
                loc: l(start),
            };
        }

        switch (text[i]) {
            case '[':
            case '{':
            case '(':
                const start = i;
                const last = pairs[text[i] as '['];
                i += 1;
                const items: RNode[] = [];
                while (true) {
                    const next = read();
                    if (!next) break;
                    items.push(next);
                }
                skipWhite();
                if (i >= text.length || text[i] !== last) {
                    throw new Error(`Expected ${last}: ${text[i]}`);
                }
                i++;
                return {
                    type:
                        last === ']'
                            ? 'array'
                            : last === ')'
                            ? 'list'
                            : 'record',
                    items,
                    loc: l(start),
                };
            case '"':
                return readString();
            case '.':
                if (text[i + 1] === '.') {
                    const loc = l(i);
                    i += 2;
                    const inner = read();
                    if (!inner) return;
                    return { type: 'spread', contents: inner, loc };
                }
            default: {
                let start = i;
                for (; i < text.length && !stops.includes(text[i]); i++) {}
                if (start === i) return;
                let idText = text.slice(start, i);
                if (idText.includes('#')) {
                    const [left, right] = idText.split('#');
                    idText = left;
                    start = +right;
                }
                let ref: IDRef;
                if (idText.includes('!')) {
                    const [left, right] = idText.split('!');
                    idText = left;
                    const key = right.length ? right : left;
                    if (left === '') {
                        ref = { type: 'placeholder', text: right };
                    } else {
                        ref = globals[key];
                        if (!globals[key]) {
                            throw new Error(`global missing: ${key}`);
                        }
                    }
                }
                const id: RNode = {
                    type: 'id',
                    text: idText,
                    loc: l(start),
                    ref,
                };
                if (text[i] === '"') {
                    const string = readString();
                    if (!string) return string;
                    return { ...string, tag: id };
                }
                return id;
            }
        }
    };
    return read();
};
