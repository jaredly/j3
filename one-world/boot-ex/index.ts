import { parse } from './format';
import { kwds } from './kwds';
import { Evaluator } from './types';

export const BootExampleEvaluator: Evaluator<void, void, void> = {
    kwds: kwds,
    parse(node, cursor) {
        return parse(node, cursor);
    },
    compile(top, info) {},
    evaluate(ir, irs) {},
    infer(top, infos) {},
    print(ir, irs) {
        return { code: '// printed', sourceMap: [] };
    },
};
