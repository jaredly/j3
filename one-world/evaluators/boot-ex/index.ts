import { parse } from './format';
import { kwds } from './kwds';
import { Evaluator } from './types';

export const BootExampleEvaluator: Evaluator<void, void, void> = {
    kwds,
    parse(node, cursor) {
        return parse(node, cursor);
    },
    macrosToExpand(node) {
        return [];
    },
    compile(top, info) {
        return { named: {} };
    },
    evaluate(ir, irs) {},
    infer(top, infos) {},
    print(ir, irs) {
        return { code: '// printed', sourceMap: [] };
    },
};
