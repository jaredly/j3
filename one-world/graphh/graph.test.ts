import { test, expect } from 'bun:test';
// so

import { reader } from '../evaluators/boot-ex/reader';
import { IDRef, Loc, mapNode, RecNode } from '../shared/nodes';

// we load .. the ... datas?

const replaceRefs =
    (refs: Record<string, IDRef>) =>
    (node: RecNode): RecNode => {
        if (node.type === 'id') {
            if (refs[node.text]) {
                return { ...node, ref: refs[node.text] };
            }
        }
        return node;
    };

test('ok', () => {
    const term1 = reader('(def x 3)', 'term1')!;
    const x1: Loc = [['term1', 5]];

    const term2 = mapNode(
        reader('(if (< x 5) "Hello" "World")', 'term2')!,
        replaceRefs({
            '<': { type: 'builtin', kind: 'value' },
            x: { type: 'toplevel', loc: x1, kind: 'value' },
        }),
    );

    expect(term1).toMatchSnapshot();
    expect(term2).toMatchSnapshot();

    // as inputs, we have {'module1': ['term1', 'term2']}
    // and ... {term1: cst, term2: cst}

    const state = {
        modules: { module1: ['term1', 'term2'] },
        moduleForTerm: { term1: 'module1', term2: 'module1' },
        terms: { term1, term2 },
    };

    /*
    procedure:
    - do a parse on everything
    - ensure that module recursion isn't happening
    - ...
    - profit

    */
});
