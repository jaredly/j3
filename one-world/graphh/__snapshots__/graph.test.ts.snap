// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ok 1`] = `
{
  "items": [
    {
      "loc": [
        [
          "term1",
          1,
        ],
      ],
      "ref": undefined,
      "text": "def",
      "type": "id",
    },
    {
      "loc": [
        [
          "term1",
          5,
        ],
      ],
      "ref": undefined,
      "text": "x",
      "type": "id",
    },
    {
      "loc": [
        [
          "term1",
          7,
        ],
      ],
      "ref": undefined,
      "text": "3",
      "type": "id",
    },
  ],
  "loc": [
    [
      "term1",
      0,
    ],
  ],
  "type": "list",
}
`;

exports[`ok 2`] = `
{
  "items": [
    {
      "loc": [
        [
          "term2",
          1,
        ],
      ],
      "ref": undefined,
      "text": "if",
      "type": "id",
    },
    {
      "items": [
        {
          "loc": [
            [
              "term2",
              5,
            ],
          ],
          "ref": {
            "kind": "value",
            "type": "builtin",
          },
          "text": "<",
          "type": "id",
        },
        {
          "loc": [
            [
              "term2",
              7,
            ],
          ],
          "ref": {
            "kind": "value",
            "loc": [
              [
                "term1",
                5,
              ],
            ],
            "type": "toplevel",
          },
          "text": "x",
          "type": "id",
        },
        {
          "loc": [
            [
              "term2",
              9,
            ],
          ],
          "ref": undefined,
          "text": "5",
          "type": "id",
        },
      ],
      "loc": [
        [
          "term2",
          4,
        ],
      ],
      "type": "list",
    },
    {
      "first": "Hello",
      "loc": [
        [
          "term2",
          12,
        ],
      ],
      "tag": {
        "loc": [
          [
            "term2",
            12,
          ],
        ],
        "text": "",
        "type": "id",
      },
      "templates": [],
      "type": "string",
    },
    {
      "first": "World",
      "loc": [
        [
          "term2",
          20,
        ],
      ],
      "tag": {
        "loc": [
          [
            "term2",
            20,
          ],
        ],
        "text": "",
        "type": "id",
      },
      "templates": [],
      "type": "string",
    },
  ],
  "loc": [
    [
      "term2",
      0,
    ],
  ],
  "type": "list",
}
`;
